### PR TITLE
[Enhancement] Optimize date_add & date_trunc

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue
1. optimize `date_add` plan：`date_add(date_add(x, 1), 2) -> date_add(x, 3)`
2. optimize `date_trunc` plan: `date_trunc('day', cast(x as date)) -> cast(x as date)`

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 42 / 44 (95.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/SimplifiedPredicateRule.java|42|44|95.45%|[358, 361]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 13 / 21 (61.90%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/cast_expr.cpp|13|21|61.90%|[824, 825, 929, 958, 959, 960, 961, 962]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 25 / 31 (80.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/cast_expr.cpp|25|31|80.65%|[851, 852, 853, 958, 959, 960]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 42 / 44 (95.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/SimplifiedPredicateRule.java|42|44|95.45%|[358, 361]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 25 / 31 (80.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/cast_expr.cpp|25|31|80.65%|[851, 852, 853, 958, 959, 960]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 25 / 31 (80.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/cast_expr.cpp|25|31|80.65%|[851, 852, 853, 958, 959, 960]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/SimplifiedPredicateRule.java [hunk 1]
```diff
@@ -315,6 +344,67 @@ public ScalarOperator visitCall(CallOperator call, ScalarOperatorRewriteContext
...
+                .findFirst().orElse("impossible");
+        if (!call.getChild(1).isConstantRef() || !Type.INT.equals(call.getChild(1).getType())) {
+            return call;
+        }
+        if (!(call.getChild(0) instanceof CallOperator)) {
+            return call;
+        }
+
+        CallOperator child = call.getChild(0).cast();
+        if (!child.getFnName().contains(fn)) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/SimplifiedPredicateRule.java:367]: may also need use `TIME_FN_NAMES.contains(call.getFnName())` to check

**Review by kangkaisen — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**